gold/Gold_data.hpp

```diff
@@ -0,0 +1,271 @@
+        c.data.emplace_back(rand_data.any());
+
+        if (chunks.size()<i && chunks[i+1].addr == addr) { //merge with next
+          c.data.insert(c.data.end(), chunks[i+1].data.begin(), chunks[i+1].data.end());
+          ++i;
+          chunks.erase(chunks.begin()+i);
+        }
+        return;
+      }
+    }
+
+    set_addr(addr, 1);
+  }
+
+  void add_addr(const uint64_t addr, const uint8_t sz) {
+    for(auto i=0u;i<sz;++i) {
+      add_addr(addr+i);
+    }
+  }
+
+  void set_data(const uint64_t addr, const uint8_t sz, uint64_t d) {
+    assert(sz<=8);
+    for(int i=0;i<sz;++i) {
+      set_byte(addr+(sz-i-1), d&0xFF);
+      d = d>>8;
+    }
+  }
+
+  uint64_t get_data(const uint64_t addr, const uint8_t sz) const {
+    uint64_t d=0;
+    assert(sz<=8);
+    for(int i=0;i<sz;++i) {
+      uint64_t b = get_byte(addr+i);
+      d = (d<<8) | b;
+    }
+
+    return d;
+  }
+
+  bool has_full_overlap(const uint64_t addr, const uint8_t sz) const {
+    for(const auto &c:chunks) {
+      if (c.addr <= addr && (c.addr+c.data.size()) >= (addr+sz)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool has_partial_overlap(const uint64_t addr, const uint8_t sz) const {
+    for(const auto &c:chunks) {
+      if (c.addr <= addr && (c.addr+c.data.size()) >= addr) {
+        return true;
+      }
+      if (c.addr <= (addr+sz) && (c.addr+c.data.size()) >= (addr+sz)) {
+        return true;
+      }
+    }
+    return false;
+  }
+  bool has_partial_overlap(const Gold_data &d2) const {
+    for(auto const &c:d2.chunks) {
+      if (has_partial_overlap(c.addr, c.data.size()))
+        return true;
+    }
+    return false;
+  }
+
+  bool has_byte(const uint64_t addr) const {
+    return has_full_overlap(addr,1);
+  }
+
+  void set_byte(const uint64_t addr, const uint8_t b) {
+    for(auto &c:chunks) {
+      if (!c.is_hit(addr))
+        continue;
+
+      c.set_byte(addr, b);
+      return;
+    }
+    assert(false); // do not call set_data/byte if addr may be invalid. Check first
+  }
+
+  uint8_t get_byte(const uint64_t addr) const {
+    for(auto &c:chunks) {
+      if (!c.is_hit(addr))
+        continue;
+
+      return c.get_byte(addr);
+    }
+
+    return rand_data.any();
+  }
+
+  bool has_data() const { return !chunks.empty(); }
+
+  void dump() const {
+    std::cout << std::hex;
+
+    for(const auto c:chunks) {
+      if (device) {
+        std::cout << " DEVICE";
+      }
+      std::cout << " data[" << c.addr << ":" << c.addr+c.data.size()-1 << "]=";
+      std::cout << std::setfill('0') << std::setw(2) << std::right << std::hex;
+      for(auto b:c.data) {
+        std::cout << static_cast<int>(b);
+      }
+    }
+  }
+
+  void add_newer(const Gold_data &d2) {
+
+    for(const auto c:d2.chunks) {
+      add_addr(c.addr, c.data.size());
+      for(auto i=0u;i<c.data.size();++i) {
+        set_byte(c.addr+i, c.data[i]);
+      }
+    }
+  }
+
+  void update_newer(const Gold_data &d2) {
```

Comment:
What's the difference between `add_newer` and `update_newer`?  Do we really need both as public API?